File: app/src/main/java/be/digitalia/fosdem/utils/ThemeManager.java
package be.digitalia.fosdem.utils;

import android.content.Context;
import android.content.SharedPreferences;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatDelegate;
import androidx.preference.PreferenceManager;

import be.digitalia.fosdem.fragments.SettingsFragment;

public class ThemeManager implements SharedPreferences.OnSharedPreferenceChangeListener {

	private static ThemeManager instance;

	private ThemeManager(@NonNull Context context) {
		final SharedPreferences defaultSharedPreferences = PreferenceManager.getDefaultSharedPreferences(context.getApplicationContext());
		updateTheme(defaultSharedPreferences);
		defaultSharedPreferences.registerOnSharedPreferenceChangeListener(this);
	}

	public static void init(Context context) {
		if (instance == null) {
			instance = new ThemeManager(context);
		}
	}

	public static ThemeManager getInstance() {
		return instance;
	}

	@Override
	public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String key) {
		if (SettingsFragment.KEY_PREF_THEME.equals(key)) {
			updateTheme(sharedPreferences);
		}
	}

	private void updateTheme(SharedPreferences sharedPreferences) {
		final String stringMode = sharedPreferences.getString(SettingsFragment.KEY_PREF_THEME, String.valueOf(AppCompatDelegate.MODE_NIGHT_FOLLOW_SYSTEM));
		AppCompatDelegate.setDefaultNightMode(Integer.parseInt(stringMode));
	}
}

File: app/src/main/java/be/digitalia/fosdem/utils/ThemeManager.kt
package be.digitalia.fosdem.utils

import androidx.appcompat.app.AppCompatDelegate
import be.digitalia.fosdem.settings.UserSettingsProvider
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.launch
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Automatically switches the light/dark theme when the user updates its preferences.
 */
@Singleton
class ThemeManager @Inject constructor(userSettingsProvider: UserSettingsProvider) {

    init {
        BackgroundWorkScope.launch {
            userSettingsProvider.theme.collect { theme ->
                AppCompatDelegate.setDefaultNightMode(theme ?: AppCompatDelegate.MODE_NIGHT_FOLLOW_SYSTEM)
            }
        }
    }
}